libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2265,6 +2265,8 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 	vector<ASTPointer<Expression const>> const& arguments = _functionCall.arguments();
 	bool argumentsArePure = true;
 
+	_functionCall.expression().annotation().calledDirectly = true;
```

Comment:
We need to distinguish between direct calls where the expression is a specific function known at compilation time:
- `foo()`
- `this.foo()`
- `c.foo()`
- `L.foo()`

and indirect ones which must go through internal dispatch because the function to be called depends on the value of the expression (which might change at runtime):
- `function() internal ptr = foo; ptr();`
- `function() external ptr = this.foo; ptr();`
- `(c.foo)()`
- `(L.foo)()`

Reply:
But wouldn't it still be a "direct call" as opposed to a "touching of the function", no matter the exact circumstances? If ignoring  the cases you list makes this value useless then sure, we should absolutely distinguish those cases...

Reply:
Depends on how you look at it. The current naming makes sense to me because a pointer introduces indirection. But anyway, that's just naming. If you can come up with a better name, I'm fine with renaming it.

The thing that matters is that the purpose of the mechanism you removed was to distinguish between calling the function in a normal way and calling it via a "pointer" that needs to be listed in the internal dispatch (rather than between calling it and just referencing it by name). So this flag should preserve that distinction if you want to use it to replace the mechanism.

Also note that tests will not necessarily fail if you get this wrong - because adding more functions than necessary to dispatch just means adding dead code and won't result in compile-time or runtime failures.